### ISSUE TITLE
Receiving Method not found: 'System.Threading.Tasks.Task`1<Microsoft.Rest.ServiceClientCredentials> Microsoft.Rest.Azure.Authentication.UserTokenProvider.LoginByDeviceCodeAsync(System.String, System.String, Microsoft.Rest.Azure.Authentication.ActiveDirectoryServiceSettings, Microsoft.IdentityModel.Clients.ActiveDirectory.TokenCache, System.Func`2<Microsoft.IdentityModel.Clients.ActiveDirectory.DeviceCodeResult,Boolean>)'.

### ISSUE THREAD

**Issue by austincrft** [Azure/azure-libraries-for-net#477]
I am trying to use the `Microsoft.Azure.Management.Fluent` library to login via Service Principal. I am using the following code [as recommended by the docs](https://docs.microsoft.com/en-us/dotnet/azure/dotnet-sdk-azure-authenticate?view=azure-dotnet#authenticate-with-token-credentials):

```
var credentials = SdkContext.AzureCredentialsFactory
    .FromServicePrincipal(
        clientId,
        clientSecret,
        tenantId,
        AzureEnvironment.AzureGlobalCloud);

var azure = Microsoft.Azure.Management.Fluent.Azure
    .Configure()
    .Authenticate(credentials)
    .WithDefaultSubscription();
```

This gives me the following error:

```
Method not found: 'System.Threading.Tasks.Task`1<Microsoft.Rest.ServiceClientCredentials> Microsoft.Rest.Azure.Authentication.UserTokenProvider.LoginByDeviceCodeAsync(System.String, System.String, Microsoft.Rest.Azure.Authentication.ActiveDirectoryServiceSettings, Microsoft.IdentityModel.Clients.ActiveDirectory.TokenCache, System.Func`2<Microsoft.IdentityModel.Clients.ActiveDirectory.DeviceCodeResult,Boolean>)'.
```

I see that this is covered in #282, which was resolved by upgrading to `1.15.1`.

Things I've tried:

- using the latest version `1.16.0`
- downgrading to `1.15.1`
- manually installing `Microsoft.Rest.ClientRuntime.Azure.Authentication` version `2.3.4`

**Comment by hovsepm** [comment]
From @rubyhog - I have the same issue. There are only two methods: CreateCredentialsFromCache / LoginByDeviceCodeAsync for Microsoft.Rest.Azure.Authentication.UserTokenProvider on version 2.3.4
Even I installed Microsoft.Azure.Management.Fluent 1.15.1 by Nuget after I saw this post, still can't see LoginSilentAsync method. ( Actaully previously 2.3.2's method names are CreateCredentialsFromCache / LoginWithPromptAsync / LoginSilentAsync )
How to use Microsoft.Azure.Management.Fluent 1.15.1 for this? thanks a lot.

**Comment by hovsepm** [comment]
@austincrft could you provide some more context on your project?

**Comment by hovsepm** [comment]
@austincrft could you also add target framework and what other nuget packages your project depends on (non fluent)?

**Comment by austincrft** [comment]
I am targeting `net461` and the only other package I have is `Autofac` version `4.6.2`

**Comment by austincrft** [comment]
In an effort to identify what is causing issues, I created a console program in the same solution with a project reference to the one that is throwing the error. 

When running that code from the console wrapper I get a different error, but on the same line.

> Could not load file or assembly 'Newtonsoft.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)

I have no references to `Newtonsoft.Json` version `6.0.0.0` in my project and the Azure library itself relies on version `9.0.1`, so I don't know how to debug this further.

**Comment by hovsepm** [comment]
@austincrft .Net 4.5.2 target has hard dependency on Newtonsoft,Json 6.0. You need to migrate to 4.6.1 or higher or to .Net Core.

**Comment by austincrft** [comment]
I am targeting `net461` as I mentioned above.

**Comment by hovsepm** [comment]
are you using Fluent 1.6.0?

**Comment by austincrft** [comment]
I am using 1.16.0, the latest version. This is my csproj.

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net461</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Autofac" Version="4.6.2" />
    <PackageReference Include="Microsoft.Azure.Management.Fluent" Version="1.16.0" />
  </ItemGroup>

</Project>
```

**Comment by hovsepm** [comment]
@can you share your sample project with [me](hovsepm@microsoft.com)?

**Comment by austincrft** [comment]
**tl;dr**: I was able to get around the issue by using `WithSubscription(subscriptionId)` instead of `WithDefaultSubscription()`

I created a sample application using the same packages, targeting the same framework, and using the same service principal. It did not throw the `Method not found` exception.

I couldn't manage to figure out why, but I found a workaround by using `WithSubscription(subscriptionId)` instead of `WithDefaultSubscription()`

The problem still exists for me if I use `WithDefaultSubscription()` but I'd be okay with closing this since the workaround is fine.

**Comment by austincrft** [comment]
Actually, that just pushed the problem further down. I am getting this same error, but on this code

```
azure.ContainerGroups.Define(groupName)
    .WithRegion(region)
    .WithExistingResourceGroup(resourceGroupName)
    .WithLinux()
    .WithPrivateImageRegistry(
        data.RegistryName,
        data.RegistryUsername,
        data.RegistryPassword)
    .WithoutVolume()
    .DefineContainerInstance(name)
        .WithImage(registryName + "/" + containerImageName)
        .WithoutPorts()
        .WithStartingCommandLine("dotnet", containerArgs)
        .Attach()
    .WithRestartPolicy(ContainerGroupRestartPolicy.Never)
    .Create();
```

**Comment by austincrft** [comment]
Btw, I still get the same error when trying to create a storage account, so the error isn't ContainerGroup-specific

```
azure.StorageAccounts.Define("solacrafttesting")
    .WithRegion("East US")
    .WithExistingResourceGroup(data.ResourceGroupName)
    .Create();
```

**Comment by hovsepm** [comment]
@austincrft I'm unable to repro. Please take a look if this is exactly what you have.
project file:
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net461</TargetFramework>
  </PropertyGroup>
  
  <ItemGroup>
    <PackageReference Include="Autofac" Version="4.6.2" />
    <PackageReference Include="Microsoft.Azure.Management.Fluent" Version="1.16.0" />
  </ItemGroup>

</Project>
```

code that does not throw any exception and runs till the end:
```
var credentials = SdkContext.AzureCredentialsFactory
    .FromServicePrincipal(
        str1,
        str2,
        str3,
        AzureEnvironment.AzureGlobalCloud);

var azure = Microsoft.Azure.Management.Fluent.Azure
    .Configure()
    .Authenticate(credentials)
    .WithDefaultSubscription();

azure.ResourceGroups.Define(rgName)
    .WithRegion(Region.USWest2)
    .Create();

azure.StorageAccounts.Define("solacrafttesting2")
    .WithRegion("East US")
    .WithExistingResourceGroup(rgName)
    .Create();

azure.ContainerGroups.Define("sdfsdsd3")
    .WithRegion(Region.USWest2)
    .WithExistingResourceGroup(rgName)
    .WithLinux()
    .WithPrivateImageRegistry(
        "asdsad",
        "asdaskdjasi",
        "asd!@#$%^456")
    .WithoutVolume()
    .DefineContainerInstance("sdfs")
        .WithImage(str4 + "/" + str5)
        .WithoutPorts()
        .WithStartingCommandLine("dotnet", "adasd")
        .Attach()
    .WithRestartPolicy(ContainerGroupRestartPolicy.Never)
    .Create();

azure.ResourceGroups.DeleteByName(rgName);
```

What is your development environment? OS + version?

**Comment by austincrft** [comment]
I'm not able to reproduce with a sample project either, unfortunately. I will have to do more work reproducing in a way that I can share. There must be something different about the projects that is causing the error.

**Comment by austincrft** [comment]
Sorry, just realized I didn't answer your question.

OS: Windows 10
Version: 1803

**Comment by hovsepm** [comment]
@rubyhog could you provide a project/code that you used to hit this issue?

**Comment by hovsepm** [comment]
Seems as non-reproducible issue. Please open a new issue with repro steps if you will hit it in v1.16.1 or higher.
